File: FileSearchMvvm/bin/Release/Models/Utilities/AcrobatJavascripts.js
/*  This file is copied from FileSearchMvvm to Acrobat's Javascript folder as:
    C:\Users\daniel\AppData\Roaming\Adobe\Acrobat\Privileged\10.0\JavaScripts\local_scripts.js
    The user acct file will only be updated if there has been a change to the original.
*/
function convertToPdfA() {
    var oProfile = Preflight.getProfileByName('Convert to PDF/A-1b (sRGB)');
    if (oProfile !== undefined) {
        var myPreflightResult = this.preflight(oProfile, false);
        return this.path;
    }
    console.println("\nPreflight profile not found.");
    return -1;
}
function getLetterSizeCoverCoordinates(cover_length) {
    if (cover_length === 48) { return [-117.5, 35, 494.5, 827]; }
    if (cover_length === 49) { return [-117.5, 30, 494.5, 822]; }
    if (cover_length === 50) { return [-117.5, 25, 494.5, 817]; }
    if (cover_length === 51) { return [-117.5, 20, 494.5, 812]; }
    else { return null; }
}
function getBookletSizeCoverCoordinates(cover_length) {
    if (cover_length === 48) { return [-32, 105.5, 409, 771.5]; }
    if (cover_length === 49) { return [-32, 98.5, 409, 764.5]; }
    if (cover_length === 50) { return [-32, 91.5, 409, 757.5]; }
    if (cover_length === 51) { return [-32, 84.5, 409, 750.5]; }
    else { return null; }
}
function getLetterSizeBriefCoordinates() { return [-117.5, 94.5, 494.5, 886.5]; }
function getBookletSizeBriefCoordinates() { return [-32, 157.5, 409, 823.5]; }
function centerPdfOnLetterPageNoCover() {
    var rectBrief = getLetterSizeBriefCoordinates();
    this.setPageBoxes({
        cBox: 'Media',
        nStart: 0,
        nEnd: this.numPages - 1,
        rBox: rectBrief
    });
    var media_box = this.getPageBox('Media');
    this.setPageBoxes({
        cBox: 'Crop',
        nStart: 0,
        nEnd: this.numPages - 1,
        rBox: media_box
    });
    return this.path;
}
function centerPdfOnLetterPageWithCover48() {
    var rectCover = getLetterSizeCoverCoordinates(48);
    this.setPageBoxes({
        cBox: 'Media',
        nStart: 0,
        nEnd: 0,
        rBox: rectCover
    });
    var media_box_cover = this.getPageBox('Media', 0);
    this.setPageBoxes({
        cBox: 'Crop',
        nStart: 0,
        nEnd: 0,
        rBox: media_box_cover
    });
    if (this.numPages > 1) {
        var rectBrief = getLetterSizeBriefCoordinates();
        this.setPageBoxes({
            cBox: 'Media',
            nStart: 1,
            nEnd: this.numPages - 1,
            rBox: rectBrief
        });
        var media_box_brief = this.getPageBox('Media', 1);
        this.setPageBoxes({
            cBox: 'Crop',
            nStart: 1,
            nEnd: this.numPages - 1,
            rBox: media_box_brief
        });
    }
    return this.path;
}
function centerPdfOnLetterPageWithCover49() {
    var rectCover = getLetterSizeCoverCoordinates(49);
    this.setPageBoxes({
        cBox: 'Media',
        nStart: 0,
        nEnd: 0,
        rBox: rectCover
    });
    var media_box_cover = this.getPageBox('Media', 0);
    this.setPageBoxes({
        cBox: 'Crop',
        nStart: 0,
        nEnd: 0,
        rBox: media_box_cover
    });
    if (this.numPages > 1) {
        var rectBrief = getLetterSizeBriefCoordinates();
        this.setPageBoxes({
            cBox: 'Media',
            nStart: 1,
            nEnd: this.numPages - 1,
            rBox: rectBrief
        });
        var media_box_brief = this.getPageBox('Media', 1);
        this.setPageBoxes({
            cBox: 'Crop',
            nStart: 1,
            nEnd: this.numPages - 1,
            rBox: media_box_brief
        });
    }
    return this.path;
}
function centerPdfOnLetterPageWithCover50() {
    var rectCover = getLetterSizeCoverCoordinates(50);
    this.setPageBoxes({
        cBox: 'Media',
        nStart: 0,
        nEnd: 0,
        rBox: rectCover
    });
    var media_box_cover = this.getPageBox('Media', 0);
    this.setPageBoxes({
        cBox: 'Crop',
        nStart: 0,
        nEnd: 0,
        rBox: media_box_cover
    });
    if (this.numPages > 1) {
        var rectBrief = getLetterSizeBriefCoordinates();
        this.setPageBoxes({
            cBox: 'Media',
            nStart: 1,
            nEnd: this.numPages - 1,
            rBox: rectBrief
        });
        var media_box_brief = this.getPageBox('Media', 1);
        this.setPageBoxes({
            cBox: 'Crop',
            nStart: 1,
            nEnd: this.numPages - 1,
            rBox: media_box_brief
        });
    }
    return this.path;
}
function centerPdfOnLetterPageWithCover51() {
    var rectCover = getLetterSizeCoverCoordinates(51);
    this.setPageBoxes({
        cBox: 'Media',
        nStart: 0,
        nEnd: 0,
        rBox: rectCover
    });
    var media_box_cover = this.getPageBox('Media', 0);
    this.setPageBoxes({
        cBox: 'Crop',
        nStart: 0,
        nEnd: 0,
        rBox: media_box_cover
    });
    if (this.numPages > 1) {
        var rectBrief = getLetterSizeBriefCoordinates();
        this.setPageBoxes({
            cBox: 'Media',
            nStart: 1,
            nEnd: this.numPages - 1,
            rBox: rectBrief
        });
        var media_box_brief = this.getPageBox('Media', 1);
        this.setPageBoxes({
            cBox: 'Crop',
            nStart: 1,
            nEnd: this.numPages - 1,
            rBox: media_box_brief
        });
    }
    return this.path;
}
function centerPdfOnBriefPageWithNoCover() {
    var rectBrief = getBookletSizeBriefCoordinates();
    this.setPageBoxes({
        cBox: 'Media',
        nStart: 0,
        nEnd: this.numPages - 1,
        rBox: rectBrief
    });
    var media_box = this.getPageBox('Media');
    this.setPageBoxes({
        cBox: 'Crop',
        nStart: 0,
        nEnd: this.numPages - 1,
        rBox: media_box
    });
    return this.path;
}
function centerPdfOnBriefPageWithCover48() {
    var rectCover = getBookletSizeCoverCoordinates(48);
    this.setPageBoxes({
        cBox: 'Media',
        nStart: 0,
        nEnd: 0,
        rBox: rectCover
    });
    var media_box_cover = this.getPageBox('Media', 0);
    this.setPageBoxes({
        cBox: 'Crop',
        nStart: 0,
        nEnd: 0,
        rBox: media_box_cover
    });
    if (this.numPages > 1) {
        var rectBrief = getBookletSizeBriefCoordinates();
        this.setPageBoxes({
            cBox: 'Media',
            nStart: 1,
            nEnd: this.numPages - 1,
            rBox: rectBrief
        });
        var media_box_brief = this.getPageBox('Media', 1);
        this.setPageBoxes({
            cBox: 'Crop',
            nStart: 1,
            nEnd: this.numPages - 1,
            rBox: media_box_brief
        });
    }
    return this.path;
}
function centerPdfOnBriefPageWithCover49() {
    var rectCover = getBookletSizeCoverCoordinates(49);
    this.setPageBoxes({
        cBox: 'Media',
        nStart: 0,
        nEnd: 0,
        rBox: rectCover
    });
    var media_box_cover = this.getPageBox('Media', 0);
    this.setPageBoxes({
        cBox: 'Crop',
        nStart: 0,
        nEnd: 0,
        rBox: media_box_cover
    });
    if (this.numPages > 1) {
        var rectBrief = getBookletSizeBriefCoordinates();
        this.setPageBoxes({
            cBox: 'Media',
            nStart: 1,
            nEnd: this.numPages - 1,
            rBox: rectBrief
        });
        var media_box_brief = this.getPageBox('Media', 1);
        this.setPageBoxes({
            cBox: 'Crop',
            nStart: 1,
            nEnd: this.numPages - 1,
            rBox: media_box_brief
        });
    }
    return this.path;
}
function centerPdfOnBriefPageWithCover50() {
    var rectCover = getBookletSizeCoverCoordinates(50);
    this.setPageBoxes({
        cBox: 'Media',
        nStart: 0,
        nEnd: 0,
        rBox: rectCover
    });
    var media_box_cover = this.getPageBox('Media', 0);
    this.setPageBoxes({
        cBox: 'Crop',
        nStart: 0,
        nEnd: 0,
        rBox: media_box_cover
    });
    if (this.numPages > 1) {
        var rectBrief = getBookletSizeBriefCoordinates();
        this.setPageBoxes({
            cBox: 'Media',
            nStart: 1,
            nEnd: this.numPages - 1,
            rBox: rectBrief
        });
        var media_box_brief = this.getPageBox('Media', 1);
        this.setPageBoxes({
            cBox: 'Crop',
            nStart: 1,
            nEnd: this.numPages - 1,
            rBox: media_box_brief
        });
    }
    return this.path;
}
function centerPdfOnBriefPageWithCover51() {
    var rectCover = getBookletSizeCoverCoordinates(51);
    this.setPageBoxes({
        cBox: 'Media',
        nStart: 0,
        nEnd: 0,
        rBox: rectCover
    });
    var media_box_cover = this.getPageBox('Media', 0);
    this.setPageBoxes({
        cBox: 'Crop',
        nStart: 0,
        nEnd: 0,
        rBox: media_box_cover
    });
    if (this.numPages > 1) {
        var rectBrief = getBookletSizeBriefCoordinates();
        this.setPageBoxes({
            cBox: 'Media',
            nStart: 1,
            nEnd: this.numPages - 1,
            rBox: rectBrief
        });
        var media_box_brief = this.getPageBox('Media', 1);
        this.setPageBoxes({
            cBox: 'Crop',
            nStart: 1,
            nEnd: this.numPages - 1,
            rBox: media_box_brief
        });
    }
    return this.path;
}

File: FileSearchMvvm/Models/Utilities/AcrobatJavascripts.cs
using System.Collections.Generic;

namespace FileSearchMvvm.Models.Utilities
{
    enum LocalJavascripts
    {
        convert_pdfa,
        center_letter_no_cover,
        center_letter_48pica,
        center_letter_49pica,
        center_letter_50pica,
        center_letter_51pica,
        center_booklet_no_cover,
        center_booklet_48pica,
        center_booklet_49pica,
        center_booklet_50pica,
        center_booklet_51pica,
        nUpCircuitCoverOn11by19,
        nUpCircuitCoverOn8pt5by23,
        arePagesSmallerThanLetter,
        arePagesLargerThanBooklet,
        centerCenteredDocOnLetterPages,
        centerCenteredDocOnBookletPages,
        extractPagesFromDocument
    }
    static class AcrobatJS
    {
        public static bool LocalJavascriptFileExists
        {
            get { return System.IO.File.Exists(localJavascriptFile); }
        }
        // for Acrobat 10
        //private static string localJavascriptFile = @"C:\Users\" + System.Environment.UserName
        //    + @"\AppData\Roaming\Adobe\Acrobat\Privileged\10.0\JavaScripts\local_scripts.js";
        // for Acrobat DC
        private static string localJavascriptFile = @"C:\Users\" + System.Environment.UserName
            + @"\AppData\Roaming\Adobe\Acrobat\Privileged\DC\JavaScripts\local_scripts.js";

        public static Dictionary<LocalJavascripts, string> Javascripts =
            new Dictionary<LocalJavascripts, string>
        {
            { LocalJavascripts.convert_pdfa, "convertToPdfA" },
            { LocalJavascripts.center_letter_no_cover, "centerPdfOnLetterPageNoCover" },
            { LocalJavascripts.center_letter_48pica, "centerPdfOnLetterPageWithCover48" },
            { LocalJavascripts.center_letter_49pica, "centerPdfOnLetterPageWithCover49" },
            { LocalJavascripts.center_letter_50pica, "centerPdfOnLetterPageWithCover50" },
            { LocalJavascripts.center_letter_51pica, "centerPdfOnLetterPageWithCover51" },
            { LocalJavascripts.center_booklet_no_cover, "centerPdfOnBriefPageWithNoCover" },
            { LocalJavascripts.center_booklet_48pica, "centerPdfOnBriefPageWithCover48" },
            { LocalJavascripts.center_booklet_49pica, "centerPdfOnBriefPageWithCover49" },
            { LocalJavascripts.center_booklet_50pica, "centerPdfOnBriefPageWithCover50" },
            { LocalJavascripts.center_booklet_51pica, "centerPdfOnBriefPageWithCover51" },
            { LocalJavascripts.nUpCircuitCoverOn11by19, "nUpCircuitCoverOn11by19" },
            { LocalJavascripts.nUpCircuitCoverOn8pt5by23, "nUpCircuitCoverOn8pt5by23" },
            { LocalJavascripts.arePagesSmallerThanLetter, "arePagesSmallerThanLetter"},
            { LocalJavascripts.arePagesLargerThanBooklet, "arePagesLargerThanBooklet"},
            { LocalJavascripts.centerCenteredDocOnLetterPages, "centerCenteredDocOnLetterPages"},
            { LocalJavascripts.centerCenteredDocOnBookletPages, "centerCenteredDocOnBookletPages"},
            { LocalJavascripts.extractPagesFromDocument, "extractPagesFromDocument"}
        };

        public static bool AreAcrobatJavascriptsInPlace()
        {
            /* Location of Javascript folder: check folder-level file
             * C:\Users\daniel\AppData\Roaming\Adobe\Acrobat\Privileged\10.0\JavaScripts\local_scripts.js;
             * Name of js function: convertToPdfA
             * Name of Acrobat Profile: Convert to PDF/A-1b (sRGB)
             */
            try
            {
                var username = System.Environment.UserName;
                // removed this fix 1/26/2023
                // if(username == "Printer") { username = "todd"; } // weird here, but necessary
                var local_script_file = @"C:\Users\" + username + @"\AppData\Roaming\Adobe\Acrobat\Privileged\10.0\JavaScripts\local_scripts.js";
                var local_script = string.Empty;
                System.Reflection.Assembly _assembly = System.Reflection.Assembly.GetExecutingAssembly();
                using(var _textStreamReader = new System.IO.StreamReader(
                    _assembly.GetManifestResourceStream("FileSearchMvvm.Models.Utilities.AcrobatJavascripts.js")))
                {
                    var js_txt = _textStreamReader.ReadToEnd();
                    local_script = js_txt;
                }
                var need_to_create_new_file = false;
                if(!System.IO.File.Exists(local_script_file))
                {
                    need_to_create_new_file = true;
                }
                else
                {
                    var current_script = System.IO.File.ReadAllText(local_script_file);
                    if(!local_script.Equals(current_script))
                    {
                        need_to_create_new_file = true;
                    }
                }
                if(need_to_create_new_file)
                {
                    System.IO.File.WriteAllText(local_script_file, local_script);
                }

                if(System.IO.File.Exists(local_script_file)) { return true; }
                else { return false; }
            }
            catch(System.Exception ex)
            {
                System.Diagnostics.Debug.WriteLine(ex.Message);
                return false;
            }
        }
        public static bool AreAcrobatSequencesInPlace()
        {
            /* Location of Javascript sequences: 
             * C:\Users\daniel\AppData\Roaming\Adobe\Acrobat\10.0\Sequences\
             * Name of centering sequence: Center Typeset Document.sequ
             */
            try
            {
                var username = System.Environment.UserName;
                var local_script_file = @"C:\Users\" + username + @"\AppData\Roaming\Adobe\Acrobat\10.0\Sequences\Center Typeset Document.sequ";
                var local_script = string.Empty;
                System.Reflection.Assembly _assembly = System.Reflection.Assembly.GetExecutingAssembly();
                using(var _textStreamReader = new System.IO.StreamReader(
                    _assembly.GetManifestResourceStream("FileSearchMvvm.Models.Utilities.Center Typeset Document.sequ")))
                {
                    var js_txt = _textStreamReader.ReadToEnd();
                    local_script = js_txt;
                }
                var need_to_create_new_file = false;
                if(!System.IO.File.Exists(local_script_file))
                {
                    need_to_create_new_file = true;
                }
                else
                {
                    var current_script = System.IO.File.ReadAllText(local_script_file);
                    if(!local_script.Equals(current_script))
                    {
                        need_to_create_new_file = true;
                    }
                }
                if(need_to_create_new_file)
                {
                    System.IO.File.WriteAllText(local_script_file, local_script);
                }

                if(System.IO.File.Exists(local_script_file)) { return true; }
                else { return false; }
            }
            catch(System.Exception ex)
            {
                System.Diagnostics.Debug.WriteLine(ex.Message);
                return false;
            }
        }
    }
}